gateway/handler_unixfs_dir.go

```diff
@@ -113,24 +117,32 @@ func (i *handler) serveDirectory(ctx context.Context, w http.ResponseWriter, r *
 		return true
 	}
 
-	results, err := i.api.LsUnixFsDir(ctx, resolvedPath)
-	if err != nil {
-		webError(w, err, http.StatusInternalServerError)
-		return false
-	}
+	var dirListing []assets.DirectoryItem
+	it := dir.Entries()
+	for it.Next() {
+		if err := it.Err(); err != nil {
+			webError(w, err, http.StatusInternalServerError)
+			return false
+		}
+
+		name := it.Name()
+		sz, err := it.Node().Size()
+		if err != nil {
+			webError(w, err, http.StatusInternalServerError)
+			return false
+		}
 
-	dirListing := make([]assets.DirectoryItem, 0, len(results))
-	for link := range results {
-		if link.Err != nil {
-			webError(w, link.Err, http.StatusInternalServerError)
+		md, ok := it.Node().(dirEntryMetadata)
```

Comment:
This is the problem flagged here https://github.com/ipfs/kubo/pull/9681#issuecomment-1463924712.

Basically we do not want to be calling `Node()` since that will actually load the block. We instead want the information in the dag-pb link (i.e. cumulative size and CID).

Some ways we can resolve this:

### A (unions in a struct)
Instead of returning a `files.Node` from `Get()` we can return something like:
```
type GetResponse struct {
    BytesResponse files.File
    DirectoryListingResponse iface.DirEntry (or something similar)
}
```
and treat it as a union where we return one of the types.

### B (type casting + go-unixfs changes)

Rewire the go-unixfs directory iterator to allow for something like `NodeMetadata() map[string]interface{}` and do a type check on that where we check for `Cid` and `CumulativeSize`

### C (type casting + new struct)

Have a new type in `files` (or just a type check in the gateway code) for `NodeMetadata() map[string]interface{}` and in blocks_gateway for directory objects return that instead of the standard `NewUnixFSFile`

----

I'll probably try out A (which will drop the need for the go-unixfs change), but leave a comment if you have a preference for one of the alternatives

Reply:
Go for it @hacdias. I'm not around until morning ET tomorrow.

Reply:
@aschmahmann I was thinking and maybe there's an option D. What about:

* Keeping the changes you've made to go-unixfs as-is. I think having the `.Cid` method can be useful.
* Create a wrapper around `files.Directory` that provides the same `files.DirListing` interface, wrapping `Links`. Basically, the old logic but wrapped in a `files.Directory` instead. 

With this we could keep a single struct, and each implementation is still free to optimize it the way they want. Wdyt?